### PR TITLE
Implement textDocument/prepareRename for better rename experience

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://github.com/dotnet/vscode-csharp/issues/6643

### Why
Two benefits of implementing prepare rename
1.  The UX is better when an item cannot be renamed (see gifs below)
2.  It allows us to play nicer with other extensions that implement rename in different contexts.

### Before this change - rename on invalid identifier
rename_fail

### After this change - rename on invalid identifier
prepare_rename

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public PrepareRenameHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(PrepareRenameParams request) => request.TextDocument;
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs:29]:
```suggestion
    public TextDocumentIdentifier GetTextDocumentIdentifier(PrepareRenameParams request)
        => request.TextDocument;
```

nit: i really prefer methods (even simple ones) having the body on the next line.  makes it immediately clear to me it's got a body.

File: src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+
+        var symbolicRenameInfo = await SymbolicRenameInfo.GetRenameInfoAsync(
+            document, position, cancellationToken).ConfigureAwait(false);
+        if (symbolicRenameInfo.IsError)
+            return null;
+
+        return new DefaultBehaviorPrepareRename
+        {
+            DefaultBehavior = true,
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs:45]: curious what 'default' means in this context.

File: src/Features/LanguageServer/ProtocolUnitTests/Rename/PrepareRenameTests.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Test.Utilities;
+using Xunit;
+using Xunit.Abstractions;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.Rename
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/ProtocolUnitTests/Rename/PrepareRenameTests.cs:13]: fsn.

File: src/Features/LanguageServer/ProtocolUnitTests/Rename/PrepareRenameTests.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+{
+    public class PrepareRenameTests : AbstractLanguageServerProtocolTests
+    {
+        public PrepareRenameTests(ITestOutputHelper testOutputHelper) : base(testOutputHelper)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/ProtocolUnitTests/Rename/PrepareRenameTests.cs:17]: nit: primary constructor.  
nit: sealed.

File: src/Features/LanguageServer/ProtocolUnitTests/Rename/PrepareRenameTests.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        {
+        }
+
+        [Theory, CombinatorialData]
+        public async Task TestPrepareRenameValidLocationAsync(bool mutatingLspWorkspace)
+        {
+            var markup =
+@"class A
+{
+    void {|caret:|}M()
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/ProtocolUnitTests/Rename/PrepareRenameTests.cs:27]: can you add a test where the caret is between the M and the ( ?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs:29]: This rule may be worth a new analyzer in Roslyn PrivateAnalyzers project.

File: src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs:42]: Default means default rename behavior.  Prepare allows you to specify a special range, but we don't use that.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs:42]: thanks!  what is an example of a 'special range'?  (just for my understanding).

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Rename/PrepareRenameHandler.cs:42]: The range to rename - see the docs on the return value here https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_prepareRename

**Comment by davidwengier** [comment]
Looked into piggy backing off this to add support to Razor, and FYI this will break rename in VS if the handler lights up there (ie, when LSP is turned on): https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1914930

**Comment by dibarbet** [comment]
@davidwengier it sounds like if I merge this it will break rename in Razor, is that correct?  Or will it continue working because Razor doesn't call Roslyn's prepare rename?

**Comment by davidwengier** [comment]
It's cool, you can merge. We don't advertise prepare rename support, so they won't call us, and we won't call you. In fact, if we do want to support we can workaround the client issue in Razor, by translating your response into a range response, so we're not blocked either way.

**Pull request merged**